### PR TITLE
colorbalancergb: orthogonal saturation and brilliance

### PR DESCRIPTION
**Pull request by flannelhead**
**DANGER: don't use in production use yet. breaks old edits**

I've been knee deep in the Color balance RGB code, trying to find optimization opportunities for the CPU path. Meanwhile I tried to gain further understanding on the saturation/brilliance/gamut mapping section of the code.

The current production code implements brilliance as a scaling of the JC vector. Saturation is implemented as a linear move in a direction orthogonal to the JC vector. As far as I understand, this is kind of a small angle approximation of what's described in the [user manual](https://docs.darktable.org/usermanual/3.6/special-topics/color-management/color-dimensions/):

> Increasing the chroma will therefore move all colors uniformly away from the central gray axis horizontally, while increasing the saturation will close or open the angle of the oblique dashed lines like a flower.

However the current algorithm doesn't preserve brilliance while changing the saturation! This is particularly observable with larger values of the adjustments. Even grey pixels will change in luminance, which is not expected.

This PR changes the saturation operation to actually interpolate the polar angle, which is to be interpreted as the saturation. This is more analogous to "spreading the dashed lines like a flower" and the norm of the JC vector, i.e. brilliance, is preserved.

I also took the opportunity to simplify the gamut mapping – it is easy enough to soft clip to the gamut limit while adjusting the saturation. The whole saturation/brilliance/gamut mapping section is more straightforward as a result.

This will, of course, break old edits. I don't want to rush this for 3.8, but this is offered as a POC for evaluation and testing. Only CPU version is done so far. If these changes are agreed upon, let's find a way to preserve the old edits while introducing the improved color science.

Pinging @aurelienpierre for initial review of the color science. The module is amazing as it is, however I believe we can make it even more so by these small improvements.

### CODE REVIEW

**Comment by aurelienpierre** [comment]
> Even grey pixels will change in luminance, which is not expected.

For grey pixels, brilliance == brightness. Brilliance is the radius that goes from pure black to the current color. 

> However the current algorithm doesn't preserve brilliance while changing the saturation!

Yes. I tried preserving brilliance (aka radius from black point) when changing the saturation and it makes dark colors clip too fast to black, compared to bright ones. I think there is even a commit that tries that and gets reverted later.

**Comment by flannelhead** [comment]
> For grey pixels, brilliance == brightness. Brilliance is the radius that goes from pure black to the current color.
I know, and indeed my assertion about grey pixels changing when changing saturation, turns out to be false at a closer look. Sorry about that...

> Yes. I tried preserving brilliance (aka radius from black point) when changing the saturation and it makes dark colors clip too fast to black, compared to bright ones. I think there is even a commit that tries that and gets reverted later.

Could you please try out these changes when time permits? This is literally coded based on what's stated in the linked color dimensions page in the manual. I skimmed through the history and didn't see anything really similar to what's done in this PR.

I don't agree about dark colors clipping to black too fast. This PR changes all pixels with the same initial saturation to the same final saturation (if we interpret saturation as the polar angle in (J, C) space, which seems to be the interpretation in the manual). With the linear offset in the orthogonal direction that is done by the current master code, this isn't really the case unfortunately.

**Comment by aurelienpierre** [comment]
I can try again, but I'm positive using the saturation as an editing parameter is too harsh, that's why I resorted to the rotation of the axes instead. I think @sobotka tried as well.

**Comment by sobotka** [comment]
Conventional definitions of “chroma” or “hue” would be rooted in an observer sensation model, not stimulus manipulation.

This is likely the reason there would be discrepancies. I reckon folks are in the wrong domain.

**Comment by flannelhead** [comment]
@sobotka this module uses the perceptually uniform JzAzBz colorspace for the saturation adjustments, which is supposed to be pretty good for this kind of stuff, and I think the success of this module also proves it in a sense.

A bit tangential, but just letting you know @aurelienpierre that I just recently started experimenting with gamut mapping the Filmic RGB output to the export profile in JzAzBz (initially just out of curiosity) - I'll post a prototype if/when there are some meaningful results. Have you yet done any experiments on that front yourself?

**Comment by sobotka** [comment]
> @sobotka this module uses the perceptually uniform JzAzBz colorspace for the saturation adjustments, which is supposed to be pretty good for this kind of stuff

Any observer sensation model is definitely a step in the right direction. There is a bit of an elephant in the room with respect to a majority of CAMs and what not, in that they all rest atop of somewhat erroneous ideas of the role of luminance stimulus.

To be clear, to distinguish between “chroma” and “saturation” here; chroma is the ratio of the adapted achromatic “spine” of an observer sensation model, and the chroma distance of the mixture in question. That is, “at equivalent brightness / lightness”.

Given that luminance was designed around the need for an _additive_ system^1, we can see the issue become more relevant to imaging.

Luminance as a simple linear gain cannot accommodate the impact of chroma’s relationship to “brightness”, which will sadly explode some expectations. The impact of a specific “angle” of chroma on observer sensation “brightness” is tremendous, and will distort any model accordingly. JzAzBz is one of those models, like the countless CAMs, that fail miserably here.

If someone is interested in connecting some of the lower level dots here, there are very few CAMs that attempt to account for this. CAM18SL for example, is worth a peek to see how they derive equivalent brightness.

TL;DR: CAMs are an excellent place to derive a ruler for measuring chroma, and inverting back to stimulus. However, plenty of the rulers are busted in rather fundamental ways with respect to image formation.

1. Herbert E. Ives (1912) XC. Studies in the photometry of lights of different colours.—IV. The addition of luminosities of different colour , Philosophical Magazine Series 6, 24:144, 845-853, DOI: 10.1080/14786441208634878

**Comment by flannelhead** [comment]
@aurelienpierre just an update and request for your opinion. I have not abandoned working on this enhancement but I've been pretty busy with other things. Still, I have sharpened the idea of what the improved saturation / brilliance algorithm could look like.

1. I would like to change the scaling of input for XYZ -> JzAzBz conversion. As described in https://github.com/darktable-org/darktable/issues/10437, JzAzBz expects absolutely scaled XYZ values i.e. representing a real luminance. In darktable that is not the case at any part of the pipeline ; rather, it would be a good idea to fix diffuse white (Y = 1) to some sensible value like 100 cd/m2 (which would be a pretty sensible display brightness in many cases).
2. Adjust brilliance first according to user settings
3. Then reduce chroma until the pixel fits in working profile gamut. This ensures the soft clamping of saturation angle in the next step doesn't unnecessarily change the brightness of otherwise unchanged pixels -> at default settings, the module's behaviour would be quite neutral.
4. Perform the saturation angle adjustment and ensure we don't go out of gamut - for that we could have a lookup table of "max saturation angle at a certain hue and brilliance". This could be probably modeled as a polynomial for each hue slice (`max_sat = polynomial function of brilliance`)

For the filmic rgb gamut mapping part, I've looked around a bit and found the following paper: [_Colour gamut mapping using vividness scale_](https://doi.org/10.2352/ISSN.2470-1173.2019.14.COLOR-079) by Baiyue Zhao, Lihao Xu and Ming Ronnier Luo. It contains a gamut compression algorithm that works in JzAzBz and could be probably used in filmic rgb to compress the display-referred result to fit in export profile gamut. This would replace the desaturation curve applied in the end of processing. The applicable part seems to boil down to the following:

1. Reduce brilliance (they call it vividness) until the color is contained within the "upper bound" of target gamut in (Cz, Jz) coordinates, i.e. the upper surface of the gamut volume.
2. If necessary, decrease Cz until the color fits within the "lower bound" of target gamut.

Both steps use soft knee clipping. The paper has nice diagrams of both steps. The biggest question for me at the moment is how to model the gamut boundaries at each hue slice. I'm thinking of something like fitting (polynomial) curves to both the upper and lower bound. The paper doesn't address that.

What do you think about this? The gamut volume modeling is the biggest part of the effort - I can try this out but I would like to hear your comments before taking on that. :)

**Comment by sobotka** [comment]
Just a note that it might be worthwhile to make sure that brilliance is appropriate with respect to what it should define.

If we are speaking in terms of the Swenholt / Evans definition of brilliance, cited by Fairchild, Heckaman, and others to follow, it is worth noting that brilliance exists along the brightness axis.

Given that a majority of the Canned Ham models like Jz and it’s related CAM ZCAM, and a plethora of others fail to properly model brightness…

Well we can expect everything derived from anything on top of them to be rather distant from any useful form.

**Comment by flannelhead** [comment]
Thank you @sobotka for the comments! Those made me actually double check a couple of terms, and turns out the "brilliance" used in the module concerned in this PR, is actually "vividness", introduced by [Berns](https://doi.org/10.1002/col.21833). It's the distance from black in a `(lightness, chroma)` coordinate system and it's precisely what's used here. The Fairchild brilliance paper is unfortunately paywalled and thus I can't find its precise definition.

JzAzBz, for example, doesn't attempt to model brightness at all but is just considered a Uniform Color Space, not a CAM. The linked [gamut mapping paper](https://doi.org/10.2352/ISSN.2470-1173.2019.14.COLOR-079) indeed used several different UCS including JzAzBz, arriving at quite good results.

But again thanks @sobotka ! Your comments have aroused my interest in all of these perceptual color attributes, and I wish to read more about them. However at this stage I would conclude that JzAzBz has proven already its usefulness in darktable for color adjustments and would probably be fit for the final export gamut mapping.

RawTherapee has also done quite nice-looking work on [color appearance modeling](https://rawpedia.rawtherapee.com/CIECAM02#Jzazbz_-_a_new_experimental_CAM_.3F) - they have a separate module for that. I'm intending to take a closer look at some point.

**Comment by sobotka** [comment]
> JzAzBz, for example, doesn't attempt to model brightness at all but is just considered a Uniform Color Space, not a CAM.

Are you sure JzAzBz doesn’t attempt to model brightness? What does Jz mean?

> RawTherapee has also done quite nice-looking work on color appearance modeling - they have a separate module for that. I'm intending to take a closer look at some point.

I am 80% confident it doesn’t work.

A _very_ easy thing to try is to open up your application of choice and:

1. Fill the canvas with 100% of a value such as middle sRGB code value 0.5.
2. Place a BT.709 blue square in the left region. Set the emission to 100%
3. Place an achromatic adapted square in the right region of the same dimensions.
4. Try to “match” the apparent brightness of the left BT.709 blue square with the achromatic patch. Make a note of the ball park code value you set.
5. Try it again with pure BT.709 red, and the rest of the primaries and pure compliments.

This is the best way to cut through the fault lines of the CAMs and models.

**Comment by flannelhead** [comment]
> Are you sure JzAzBz doesn’t attempt to model brightness? What does Jz mean?

It's _lightness_, I believe. The [ZCAM paper](https://doi.org/10.1364/OE.413659) defines it like this:
> brightness of an area judged relative to the brightness of a similarly illuminated area that appears to be white or highly transmitting
i.e. it's a scale relative to a reference white (which is that of the SL2 training data set, 997 Cd/m2, as far as I correctly read the paper). As far as I understand, brightness is an absolute scale instead.

It probably comes across that I'm quite a newbie at perceptual color appearance models, so I may still understand some things wrong... However, I believe that due to its hue linearity properties (under change of chroma and lightness), JzAzBz should be very much applicable for things like creative color grading (as is used here) and gamut mapping. But I'm reading the references you posted with great curiosity and experimenting to understand things better.

**Comment by sobotka** [comment]
The typical division between brightness and lightness is that lightness is a higher order, and vastly more complex surface. It’s a relative judgement anchored in brightness, and very, very complex.

The key point I would encourage your, or anyone, with an interest in forming an image from a stimulus encoding, is to perform that trivial at-home test.

Jz is indeed directly related to brightness in the model, but because of the catastrophic nature of the base assumptions, even a dilettante hobby person can discover the issue on their own via such a test above.

If you perform the test, jot down the numbers you set the “value” or “lightness” (whatever the label is in your DCC of choice) and note them here. I’ll do my best to illustrate in plain language just how crazy the results are, and you can infer on your own whether or not it would have an impact on chroma, colourfulness, brightness, “vivid”, etc.

**Comment by flannelhead** [comment]
This is educational indeed. The results I ended up with were something like:

```
    ((1., 0., 0.), 0.8),
    ((0., 1., 0.), 0.95),
    ((0., 0., 1.), 0.7),
    ((1., 0., 1.), 0.8),
    ((1., 1., 0.), 1.),
    ((0., 1., 1.), 1.),
```
where first is the RGB tuple (sRGB) and second is the L value of HSL = (0, 0, L) which is then converted to a RGB. Truth to be told, for the patches with G = 1, especially yellow, I found the patch roughly equally bright, if not brighter than the achromatic patch with L = 100%.

On the JzAzBz side for these pairs, it seems the achromatic one generally has greater Jz than the colourful one, and the ratio varies between patches. The biggest discrepancies seem to be with the patches not containing green. (The white luminance Y in this case was fixed to 100 Cd/m2.)

```
rgb = (1.0, 0.0, 0.0), l = 0.80; color Jz = 0.099, grey Jz = 0.136
rgb = (0.0, 1.0, 0.0), l = 0.95; color Jz = 0.132, grey Jz = 0.159
rgb = (0.0, 0.0, 1.0), l = 0.70; color Jz = 0.069, grey Jz = 0.119
rgb = (1.0, 0.0, 1.0), l = 0.80; color Jz = 0.118, grey Jz = 0.136
rgb = (1.0, 1.0, 0.0), l = 1.00; color Jz = 0.157, grey Jz = 0.167
rgb = (0.0, 1.0, 1.0), l = 1.00; color Jz = 0.144, grey Jz = 0.167
```

**Comment by sobotka** [comment]
> Truth to be told, for the patches with G = 1, especially yellow, I found the patch roughly equally bright, if not brighter than the achromatic patch with L = 100%.

Would you be shocked if I told you this is absolutely congruent with existing research dating as far back as 1890?

Now you seem very capable. Convert the target value and your adjusted brightness swatch into luminance using the RGB code values output. Remember you'll need to convert them to a linear stimulus encoding, so if they are expressed in RGB, take the RGB values and normalize 0.0 to 1.0, then roll the result through the assumed EOTF of your display of 2.2.

How do the luminances of the target match to the calculated adjusted you dialled in?

For even greater insight, lower the target swatch by 10% or so, and see what happens.

Excellent work!

**Comment by flannelhead** [comment]
> How do the luminances of the target match to the calculated adjusted you dialled in?

For example, the blue patch seems to have quite a lower luminance (Y) than the "equivalent brightness" grey. 

I've done my homework and learned it is the [Helmholtz-Kohlrausch effect](https://en.m.wikipedia.org/wiki/Helmholtz%E2%80%93Kohlrausch_effect) at play here. And indeed, reading the CAM18sl paper, it is one of the few CAMs that take the effect in account.

I can also see the consequences in saturation adjustment. If we apply such an adjustment in a color space not taking HK effect in account, different hues seem to act in a pretty different manner. Starting from a red, blue and green of approximately equivalent brightness and then increasing the saturation (increasing the `colorfulness / brightness` ratio by some means), the blues seem to lose brightness way faster than other colors. For the gamut compression idea discussed here it would not matter as much because there we're talking mostly about very small adjustments to bring the color from a wider source space (bounded) to a smaller one.

As you @sobotka are very knowledgeable about these things, are you aware of any color space (or CAM) that would meet the requirements of hue linearity (required for color grading applications) and taking the HK effect in account, i.e. something like JzAzBz but accounting also for the HK effect? Also in darktable's scene-referred workflow, it would be required that the model works well for HDR i.e. very high contrast ratios.

CAM18sl seems promising but it seems to target a kind of a specific scenario of a self-luminous small stimulus on a self-luminous background (although I could see this kind of applying to to photo editing scenarios). It would be probably worthfile to experiment with it. I know there's also CAM20u as an extension to CAM16, but I haven't had a chance to look at it since the paper is paywalled.

Again thanks for taking the time to give this very valuable lesson! I was actually amazed how easy the effect is to reproduce qualitatively.

**Comment by sobotka** [comment]
> I can also see the consequences in saturation adjustment.

If we were to be bold, we could extrapolate this rather significant component to an entire model issue.

For example, if we try to arrive at “workable” constructs of the problem, we could say that we have a CIE tristimulus encoding domain, let’s call it C. One such division is a three dimensional Cartesian system, which allows us to arrive at a cylindrical representation:
```
C_luminance -> Luminance
C_angle -> Chromaticity Angle
C_distance -> Chromaticity Distance
```

For the purposes of image formation and work, we can see a hint of a parallel “first order” observer sensation domain, which we can call S:
```
S_brightness -> Brightness
S_angle -> Hue
S_distance -> Chroma
```

Both domains are potentially interlinked, but nonuniform with respect to one another. In an ideal situation, we'd have the earliest Helmholtz / Sinden^1 hypotheses. Let prime denote compliment:
```
S_Brightness S_Angle S_Distance + S_Brightness' S_Angle' S_Distance' = 0.0
```

We can quickly see how if we bog up the first principle of Brightness, all of Saturation / Colourfulness, Chroma, etc. becomes mangled, which is the sole reason I brought this up in this commit. Use whatever works pragmatically, but don't place too much faith on this era of computer fit models to random abstract data. The first principles might be more important in an era where a MacBook Pro has a primary near spectrally pure!

> I've done my homework and learned it is the Helmholtz-Kohlrausch effect at play here. And indeed, reading the CAM18sl paper, it is one of the few CAMs that take the effect in account.

Indeed. I'd suggest that it is prudent to realize there are several intertwined effects that we can predict based on the above speculation. Not the least of which is perhaps one of the more poorly understood or overlooked facet of boundary conditions, specifically the "greyness" boundary^2^3.

> For the gamut compression idea discussed here it would not matter as much because there we're talking mostly about very small adjustments to bring the color from a wider source space (bounded) to a smaller one.

Plenty of nuance!

Also, have your rolled smooth and pure chroma gradient sweeps through your attempts? What you might find is that there's a profound impact on the reprojections on tonal smoothness in the smaller footprint. It's worth rolling up some unit tests that perform smooth sweeps and see how they end up on the output side.

> As you @sobotka are very knowledgeable about these things, are you aware of any color space (or CAM) that would meet the requirements of hue linearity (required for color grading applications) and taking the HK effect in account, i.e. something like JzAzBz but accounting also for the HK effect? Also in darktable's scene-referred workflow, it would be required that the model works well for HDR i.e. very high contrast ratios.

I'm just another bumbling astronaut floating in space! Sometimes we bump into other astronauts!

What I can say with almost _certainty_ is that the short answer is _no_. Most models are fit, and at that point, they explode.

Some things that might be worth considering, again in relationship to the above "first order" model consideration:
1. There's some agreement that the basic cone saturation model is something akin to a Michelis Menten shape^4^5.
2. The medium plays an absolutely fundamental role here. My personal vantage is that the act of image formation is radically different than merely approaching this from the stimulus encoding side; mediums are limited and the representation of an image deviates from those "source" signals in a way that is tightly wound up with observer visual sensation^6.
3. On the potential link between first order brightness and luminance, we can see that this has been addressed in the past, and literally every single computer display we have ever looked at. The apparent brightness is what allows us to balance to achromatic in the deeper electrical engineering side.
4. Following 3., some folks have made some connections to this with respect to brightness sensation^7.
5. There is a clear connection also to saturation^8.

Hopefully this tickles some inspiration nerves. Keep grinding!

----
1. "Studies Based on the Spectral Complementaries", R. H. Sinden.
2. "Chromatic Strength of Colors I: Dominant Wavelength and Purity", Bonnie Swenholt, Ralph Evans.
3. http://www.huevaluechroma.com/011.php, David Briggs.
4. "S-Potential and Dark Adaptation in Fish", Kishida Naka, William Rushton.
5. "The Computation of Lightness by the Primate Retina", David Marr.
6. "Quality of Color Reproduction", David MacAdam.
7. "Model of Saturation and Brightness: Relations with Luminance", Ralph Pridmore.
8. "Some Quantitative Aspects of an Opponent-Colors Theory. I. Chromatic Responses and Spectral Saturation", Dorothea Jameson, Leo Hurvich.

**Comment by sobotka** [comment]
I loathe multiple comments, however in this case, I feel it is important to pull this out in isolation.

> Also in darktable's scene-referred workflow, it would be required that the model works well for HDR i.e. very high contrast ratios.

I have always had a rather important distinction between the open domain "light like" encoding and the image formed, connected to point 2. above. I would not be so hell bent on open domain encodings as being the end all. We'd end up swapping out one Chrome Calf for another. There are also some pipelines that help to illustrated that it is nonsensical to try to do everything on the open domain encodings! See AA, upsampling / reconstruction, and many others. If this sounds rather too abstract, consider black and white film, and what you'd want to resample / reconstruct from? The open domain stimulus that forms the "extremely bright blue" in the open domain, or the closed domain image stimulus that is some level of grey, rolled through complex image formation processes? That's the basics of AA, filtering, and many other things that hint at the idea that the closed domain image stimulus encoding is something very different to open domain stimulus. Deconvolution versus sharpening is another clear division.

What I _think_, and again I'm just a bumbling astronaut here, is that our history of image formation in digital has a history that loosely is:
1. From film to electronic. First was scanning of the "image".
2. From monochromatic to chroma displays.
3. Doing math on monochromatic assumptions. See also confusion over "tone curves" and the above discussion, as well as ideas around "exposure". (EG: Exposure is luminance. [Spot meter a blue swatch versus a green](https://www.youtube.com/watch?v=0iBycrvY91I) under identical illumination levels and you'll find your exposure is different!)
4. Doing math on encoded values instead of their represented emission quantities. ("Linear light")
5. Seismic shift to working on open domain encodings, and deferring the View encoding to further down. (EG: Nuke)
6. Serious problems with this. See also the never-ending shit show known as ACES, and other problems of attempting to shoe-horn everything into open domain encoding manipulations without fully respecting the output medium.

I would probably suggest that we might see value if we focus on the output mediums, and how we might privilege specific facets of the standard observer visual system to form the image. For example, how do we negotiate "contrast" between chroma and brightness with a medium that is fundamentally limited and cannot express what is indicated in the open domain encoding?

TL;DR: I've grown very partial to dividing up the problem into two broad domains.

1. The open domain stimulus encoding. EG: "The stuff in front of a virtual camera".
2. The closed domain image encoding. EG: "The formation of the stimulus into an image via film".

I _do not_ currently believe, and also firmly believe there is plenty of evidence to support the claim as per the above examples, that we can tackle literally everything in the open domain encoding. It's foolish. I also firmly believe that we cannot address all of the issues in the closed domain image formed. I would extend that to then include manipulations on either side of that fence.

Some processes must happen on the open domain stimulus to be closer to correct with respect to our models, and some processes must happen on the closed domain, fully formed image.

Sorry for this little side bar.

**Comment by aurelienpierre** [comment]
@flannelhead I found yesterday the possible reason for discrepancies in saturation intensity : https://bottosson.github.io/posts/oklab/#munsell-data

image

JzAzBz has uniform hue but non-uniform chroma. Applying a gain on this (even through the saturation) may not be well behaved.

I will test today the arctan method with Oklab space:

image

**Comment by flannelhead** [comment]
Interesting. Oklab seems to also have quite uniform lightness: https://raphlinus.github.io/color/2021/01/18/oklab-critique.html

Please let me know of the results of the experiment. :) From my previous experiments with JzAzBz, I think we would really also need a more accurate gamut map in this module, as a function of hue and lightness. I'm not sure how is the case with Oklab. Also one reason for blues getting too dark is that the go quickly out of the display color space gamut and filmic at the end doesn't catch it all. But yes, Oklab seems worth trying.

As far as I understand the lightness estimation isn't optimized for HDR, but I wonder how much it matters since the separation of shadows / midtones / highlights seems to still take place in a limited range. Probably if one wishes to grade specular highlights separately, then that would become more relevant.

**Comment by sobotka** [comment]
> JzAzBz has uniform hue but non-uniform chroma. 

Given that chroma is directly related to brightness, how could either work noting that neither accounts for brightness?

Unless a model properly gains each hue angle, I am unsure that any attempt to derive an HSL type model would be close?

**Comment by aurelienpierre** [comment]
> Unless a model “properly” gains each hue angle, I am unsure that any attempt to derive an HSL type model would be close?

Well, the main reason for using saturation (here expressed as Cz / Jz) was to be able to degrade red to pink, and all other colors to pastels tones, in a smooth way. So, unless you have a real brightness-based model to propose, JzAzBz has proven itself more than fair-enough, and Oklab seems an even better fit.

**Comment by aurelienpierre** [comment]
> From my previous experiments with JzAzBz, I think we would really also need a more accurate gamut map in this module, as a function of hue and lightness. 

That would be a 2D LUT, and working between 0 and 10.000 Cd/m² is not a good candidate for that. The bottom part of the gamut is very close to a cone, so a 1D LUT of chroma/lightness = f(hue) seems good enough.

**Comment by sobotka** [comment]
If saturation is the long diagonal between blackness and the swatch in question, I can only think it’s totally accidental.

I would consider experimenting with Nayatani, or a lower level correction first. Perhaps there could be some tendrils worth following up via CAM18SL or something else that more properly corrects the brightness scale?

I am all for “Good Enough” as you well know. I tend to believe that the magnitude of the discrepancy here is a recipe for low hanging fruit of “better”.

At least worth taking a peek, given the implications across a large surface of other areas?

**Comment by aurelienpierre** [comment]
Tested this method in OKLab, and the results are more uniform over hues indeed.

**Comment by aurelienpierre** [comment]
But indeed, the direct `atan` method does not degrade red to pink, it barely brightens. So it is actually much closer to a direct chroma change.

**Comment by sobotka** [comment]
Give it about one month. A much wiser fit model will be available then.
